### PR TITLE
GHC 8.0 - 8.4 compatibility

### PR DESCRIPTION
**Pull request by apoikos**
This PR aims at achieving single-source compatibility with GHC 8.0 - 8.4 (LTS 7 through LTS 12). It contains some of @bsrk's work for GHC 7.10 compat from `attic/master` and a refined form of the patches we had been carrying in the Debian package since Stretch. On top of that, I'm also dropping support for ancient versions of libraries we no longer need to care about.

After merging this, the minimum supported GHC version will be 8.0/base-4.9.0.

This PR addresses #1312 #1222 #1287 and #1313. I'll need to reference the issues from the respective commits as well.

### CODE REVIEW

**Comment by iustin** [comment]
@apoikos - until the CLA is removed, do you know what's the CLA status for the few commits which are unclear?

**Comment by apoikos** [comment]
@iustin: the commits by @bsrk were already on attic/master. The snap-server 1.x support patch was contributed by @yiannist who at the time was working at GRNET and had already signed a CLA (@yiannist has a bunch of commits on master).

File: src/Ganeti/THH.hs [hunk 1]
```diff
@@ -1203,8 +1206,13 @@ genDictObject :: (Name -> Field -> Q Exp)  -- ^ a saving function
               -> Q [Dec]
 genDictObject save_fn load_fn sname fields = do
   let name = mkName sname
+      -- newName fails in ghc 7.10 when used on keywords
```
[the remark this answers is not in the document]
Comment (apoikos) [reply src/Ganeti/THH.hs]: Apparently yes: https://ghc.haskell.org/trac/ghc/ticket/10599

**Comment by iustin** [comment]
OK, setting cla to yes.

**Comment by iustin** [comment]
Feel free to merge at your leisure.

**Pull request merged**